Provide getProperties method on Configuration that does not filter

dlmarion opened the pull request:
The getProperties method which takes a filter is inefficient when
you know that you want to get all properties or you know the names
of the properties. Provide a mechanism to get the properties w/out
using a filter

EdColeman commented on core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java:
```diff
@@ -71,17 +70,13 @@ private static class PrefixProps {
-   * <p>
-   * Note: this is inefficient, but convenient on occasion. For retrieving multiple properties, use
-   * {@link #getProperties(Map, Predicate)} with a custom filter.
-   *
```
The comment that this is inefficient for multiple properties still applies, the advice to use filter maybe should be adjusted to the new methods, but in general prefer getting multiple properties so the "read" to get the underlying configs is minimized.

dlmarion replied:
> The comment that this is inefficient for multiple properties still applies

Yeah, that's fair. I can put the comment back into the javadoc for that method.

dlmarion replied:
javadoc reverted in 300afea.

EdColeman commented on core/src/test/java/org/apache/accumulo/core/conf/AccumuloConfigurationTest.java:
```diff
@@ -173,6 +173,19 @@ public String get(Property property) {
       return v;
     }
 
+    @Override
```
Is it necessary to duplicate the function here instead of delegating to the method in AccumuloConfiguration? Anything using this method is only valid as long as both stay in sync?  (I think this applies to the other methods that that have a concrete implementation in AccumuloConfiguration

dlmarion replied:
> Is it necessary to duplicate the function here instead of delegating to the method in AccumuloConfiguration?

Yes, I believe so. The methods in AccumuloConfiguration are abstract, there is no implementation. The version of getProperties that takes the filter is also abstract, so it's duplicated in all of the subclasses as well.

EdColeman reviewed (commented):
Overall these seems like a good change.

In general:
- it seems that the same code is added multiple places - can any of these be consolidated?
- In tests, seems like there could be places to test the actual code rather than providing duplicate functions in test classes.
- (maybe separate issue)  there seems to be parallel handling of Properties and Strings and no clear distinction when to use one or the other. I think the Strings were provided as a convenience and a bridge from user-land. There may be opportunities to provide the methods operating on properties and the have string convenience methods convert to a properties and then delegate to the property methods for the lookup / work. That way the functionally is implemented once and any String convenience methods just translate.  It may be desirable to handle that translation higher / earlier in the stack and not provide String lookup methods at all.

dlmarion commented:
I agree with your observations here. I'll admit that I don't fully understand the layers of configuration objects here. I do think there is an opportunity to consolidate some of this logic, but that's a different PR possibly for another time.

ctubbsii reviewed (commented):
What was the point of this added API? There's nothing that uses it, and AccumuloConfiguration is an internal class. If this isn't going to be used by any code, then why add it?

dlmarion commented:
It's being used in `AccumuloConfiguration.get(String)` and `AccumuloConfiguration.iterator`

ctubbsii commented:
> It's being used in `AccumuloConfiguration.get(String)` and `AccumuloConfiguration.iterator`

That was a big code change for such a small convenience, I think. It doesn't save from having to iterate over the properties in `AccumuloConfiguration.get(String)` case, either. And, in the `AccumuloConfiguration.iterator` case, I think there's a bug, because `getProperties(entries)` there now seems equivalent to `getProperties(entries, x -> false);` instead of `getProperties(entries, x -> true);`.

In the best case scenario, it seems we're avoiding constructing a Predicate in two places in internal code in favor of constructing a varargs arrays instead. I'm not sure that's better in terms of performance, and there's a lot more code to get the job done. Overall, I don't think this change was better than what was there before.

dlmarion commented:
>  It doesn't save from having to iterate over the properties in AccumuloConfiguration.get(String) case, either (because we're still iterating in the new method body)

We are iterating over the varargs method parameter and pulling the property value from the map of properties vs testing each entry in the map of properties to see if the Predicate returns true.

> in the AccumuloConfiguration.iterator case, I think there's a bug, because getProperties(entries) there now seems equivalent to getProperties(entries, x -> false); instead of getProperties(entries, x -> true);, because the varargs is empty (meaning no properties are specified and matched, instead of all of them matching).

I handled this in DefaultConfiguration.getInstance, but I think you are right in that I missed this in the other classes, I will submit a follow-on to fix this.

> In the best case scenario, it seems we're avoiding constructing a Predicate in two places in internal code in favor of constructing a varargs arrays instead. I'm not sure that's better in terms of performance, and there's a lot more code to get the job done. Overall, I don't think this change was better than what was there before.

I disagree. In the case where the caller wants to get the value for a single property (or even a few) with fully qualified names its certainly going to be more efficient to pull the properties from the map by key vs iterating over the entire map to test each key.

ctubbsii commented:
> We are iterating over the varargs method parameter and pulling the property value from the map of properties vs testing each entry in the map of properties to see if the Predicate returns true.

Oh right. Still, this is a small scalar multiple difference. I don't think it will substantially affect performance, given that looking up a small number of properties this way would be rare anyway.

> I disagree. In the case where the caller wants to get the value for a single property (or even a few) with fully qualified names its certainly going to be more efficient to pull the properties from the map by key vs iterating over the entire map to test each key.

Given your correction to my misunderstanding above, I agree with you. However, I don't think it matters that much in terms of performance. This is internal code, and it's unlikely we're hitting this often enough to need to optimize this specific case, especially since the caller can just do (pseudo-code): `for x in A,B,C { conf.get(x) }`

dlmarion commented:
> I don't think it will substantially affect performance, given that looking up a small number of properties this way would be rare anyway.

> However, I don't think it matters that much in terms of performance

Well, I would point you to this comment above (https://github.com/apache/accumulo/pull/2834#issuecomment-1198394023). Anything that uses `AccumuloConfiguration.get(String)` is more efficient with this code. There are several places where get(String) is being used in the internal code. Here's one good example:

  RFileOperations.openWriter() -> TableConfiguration.get(Property) -> NamespaceConfiguration.get(Property) -> AccumuloConfiguration.get(String).

ctubbsii commented:
> > I don't think it will substantially affect performance, given that looking up a small number of properties this way would be rare anyway.
> 
> > However, I don't think it matters that much in terms of performance
> 
> Well, I would point you to this comment above ([#2834 (comment)](https://github.com/apache/accumulo/pull/2834#issuecomment-1198394023)). Anything that uses `AccumuloConfiguration.get(String)` is more efficient with this code. There are several places where get(String) is being used in the internal code. Here's one good example:

That comment reveals an entirely different issue... that is a bug, where the call to Namespace.get(Property) delegates to it's parent using `getParent().get(String)` instead of `getParent().get(Property)`, like it should.

Retrieving by Property should be very efficient. Retrieving by arbitrary String should be rare, and almost never used. I believe you've optimized for a case that shouldn't be called instead of fixing the calling code.

dlmarion commented:
> I believe you've optimized for a case that shouldn't be called instead of fixing the calling code.

That's a fair observation. I didn't question the existing code. It may very well be that the existing code can be fixed and these two commits reverted.

dlmarion commented:
@ctubbsii - are you working on anything related to this? If not, I can start on the things you mentioned above.

ctubbsii commented:
@dlmarion wrote:
> @ctubbsii - are you working on anything related to this? If not, I can start on the things you mentioned above.

I haven't started anything yet. Feel free to look into it. I was going to fix the calling code, and then check to see if there were any other callers that needed to be similarly fixed, before even considering whether to revert these two commits. I did notice a test failure in NamespacesIT, after the second commit that was related to configuration property inheritance, and I think is related that needs to be investigated as well.

dlmarion commented:
@ctubbsii  - See https://github.com/apache/accumulo/pull/2864. I reverted the two commits and added a 3rd. In the 3rd commit I reduced calls to AccumuloConfiguration.get(String) and optimized that method.
